iocore/net/BIO_fastopen.cc

```diff
@@ -27,13 +27,26 @@
 
 #include "BIO_fastopen.h"
 
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#define BIO_set_data(a, ptr) ((a)->ptr = (ptr))
+#define BIO_get_data(a) ((a)->ptr)
+#define BIO_set_init(a, init) ((a)->init = (init))
+#define BIO_set_shutdown(a, shut) ((a)->shutdown = (shut))
+#define BIO_get_shutdown(a) ((a)->shutdown)
+#endif
+
+struct Data {
+  Data() : fd(NO_FD), dst(nullptr) {}
+  int fd;
+  const sockaddr *dst;
+};
```

Comment:
I think that we can work around needing a separate `Data` struct. Since we actually have a socket we can now directly hook the `ctrl` method from the socket `BIO`:

```C
const BIO_METHOD * s = BIO_s_socket();
BIO_meth_set_ctrl(fastopen.method, BIO_meth_get_ctrl(s));
```
We can set the `ptr` field with `BIO_set_data`.  We can use `BIO_get_fd` to get the file descriptor when we need it.